### ISSUE TITLE
Enable auto-switching between yaml and ansible language

### ISSUE THREAD

**Issue by ssbarnea** [ansible/vscode-ansible#582]
### Problem

As one of the most common confusions with vscode extension is the fact that vscode often fails to decide which document language to use between generic `yaml` and `ansible` language. That is because at this moment only filename patterns are use to determine if a file uses one language or another. Once user manually change the language the IDE remembers the selection, but our goal is to avoid the need for this manual selection.

With @ganeshrn help, we identified several existing extensions that hook into document open event and decide to change the language of the document based on the file content, such the shebang line for example:
* https://marketplace.visualstudio.com/items?itemName=davidhewitt.shebang-language-associator (updated 2022)
* https://marketplace.visualstudio.com/items?itemName=ryzngard.vscode-header-source 
(updated 2019)

### Solution

We need to implement the same hook for document open even and when the current document is using either `yaml` or `ansible`, use a detection function that is guessing if the file is a playbook or a taskfile.

We currently have the `isPlaybook` function but we need a new one named `detectDocumentLanguage` which receives the open document and returns the name of the vscode language that vscode should use for it.

The function should return `null` if we were not able to make a decision. For example if current document is not YAML or the YAML loading fails, we are expected to return null, as we cannot make a decision.

Once the document is loaded, the logic should be:
- if current document language is not already `ansible` or `yaml`, return `null`.
- if document is `null`, empty list, string or empty mapping, return `null` (inconclusive)
- if document is a non empty mapping, return `yaml`
- if document is a non empty sequence, check if the first element is a dictionary having either one of the following keys: `name`, `hosts`, `import_playbook`, return `ansible`.
- return `null` as fallback, as we do not want to auto-switch document type if we are not confident enough.

### Additional context

Please look at current extensions that do this switching as all of them are open source and the implementation for document open is quite easy.

This feature does not require us to touch ansible-language-server project, as this hook is specific to vscode client.

**Comment by bcoca** [comment]
#4, don't check for name, it is both not required and not deterministic of being a playbook

A problem does arise when trying to detect task files, like a role's tasks/main.yml, it is still a list, but the only requirement for a task is to have an action, which can be `action`, `local_action` or `module/action plugin name`.

**Comment by ssbarnea** [comment]
In fact we do not have a problem with tasks because we always had a requirement for tasks file to be under tasks/ folder. That was the easy one.... unless user made the unfortunate decision not to put them there.

**Comment by bcoca** [comment]
it is not a requirement for include_tasks (and up to recent versions of ansilbe, not a requirement for roles either)

**Comment by ssbarnea** [comment]
Yeah, not required. But when someone shots himself in the foot, you cannot blame the weapon. In fact ansible can also load files with any extensions, yml or yaml not being required, but that does not mean we can support anything that happens to work.

Once we fix the auto-detection users will be VERY happy!

**Comment by ganeshrn** [comment]
@bcoca  
>  don't check for name, it is both not required and not deterministic of being a playbook

Checking the name is not required in case the playbook is already written and valid. In the case of the editor when the user installs the ansible extension it is expected that the new file created with `.yml`/`.yaml` is identified with ansible language similar to how the editor associates a `.py` extension automatically to `python` language without any modelines or file association settings. 

<img width="1087" alt="Screen Shot 2022-08-05 at 8 04 41 AM" src="https://user-images.githubusercontent.com/5292461/182989881-240d3528-cd07-406c-ad7b-c46d35869887.png">

**Comment by ganeshrn** [comment]
@ssbarnea 

> if document is null, empty list, string or empty mapping, return null (inconclusive)

This is where the double extension can help. If the document has double extension return `ansible` language by default. imo the order of checking can be
* If double extension present return `ansible` language
* If modeline present for `ansible` return `ansible` language
* If the document is a non-empty sequence, check if the first element is a dictionary having either one of the following keys: name, hosts, import_playbook, return ansible. 
 * If not move on to introspect the contents and file paths.
 * If none of the above conditions passes and if the file extension ends with `.yml`/`yaml` return `YAML` language else return `null`

**Comment by ssbarnea** [comment]
Double extension should be something to implement as a later improvement, I did not mention it here because I wanted to avoid increasing complexity of this feature.

**Comment by ganeshrn** [comment]
@ssbarnea I don't get what's the complexity with double extensions.

**Comment by ssbarnea** [comment]
I marked this task as "good first issue" in order to make it ease to implement by a new contributor. I have nothing against extensions, it is just that they can be fully separated from the initial implementation. Practically I tried to describe a [MVP](https://en.wikipedia.org/wiki/Minimum_viable_product) for the first version of the auto-detection.

Same thing applies for the shebang/modeline, which can also be used, but again they can safely be done separately.

**Comment by bcoca** [comment]
You don't need to do everything in one PR, it is normally better to do smaller incremental changes and have those tested/used before adding to them.

**Comment by MarkusTeufelberger** [comment]
Would it be possible to apply the schemata for ansible content to y(a)ml files as a strong hint about their "language"? A file that doesn't match them _could_ still be an ansible file, but ones that match _very_ likely are true positives.

So far it still seems like a file such as `roles/tasks/main.yml` will likely be miscategorized as `yaml` as it doesn't refer to `hosts` or `import_playbook`.

Would a patch to extend https://github.com/ansible/vscode-ansible/blob/038b856fe98cab77adc7f73ca728d93474c29cd4/package.json#L297-L301 be welcome that at least adds `"**/tasks/*.y(a)ml"` and `"**/handlers/*.y(a)ml"`?

**Comment by IceWreck** [comment]
I just added

```json
    "files.associations": {
        "**/roles/**/tasks/*.yml": "ansible",
        "**/roles/**/vars/*.yml": "ansible",
        "**/roles/**/defaults/*.yml": "ansible",
        "**/playbooks/**/*.yml": "ansible",
    }    
```

to seetings.json

**Comment by dcd-arnold** [comment]
> I just added
> 
> ```json
>     "files.associations": {
>         "**/roles/**/tasks/*.yml": "ansible",
>         "**/roles/**/vars/*.yml": "ansible",
>         "**/roles/**/defaults/*.yml": "ansible",
>         "**/playbooks/**/*.yml": "ansible",
>     }    
> ```
> 
> to settings.json

These may be helpful, too:
```jasonc
        "**/roles/**/handlers/*.yml": "ansible",
        "**/roles/**/meta/*.yml": "ansible",
        "**/group_vars/*.yml": "ansible",
        "**/host_vars/*.yml": "ansible",
```

- https://docs.ansible.com/ansible/latest/tips_tricks/sample_setup.html#sample-ansible-setup
